src/sentry/integrations/gitlab/webhooks.py

```diff
@@ -79,92 +136,68 @@ def create_commits(self, event, organization, repo):
-
         return super(GitlabWebhookEndpoint, self).dispatch(request, *args, **kwargs)
 
-    def post(self, request, organization_id):
+    def post(self, request):
         try:
-            organization = Organization.objects.get_from_cache(
-                id=organization_id,
-            )
-        except Organization.DoesNotExist:
-            logger.error(
-                PROVIDER_NAME + '.webhook.invalid-organization',
+            integration = Integration.objects.filter(
+                provider=self.provider,
+                external_id=request.META['HTTP_X_GITLAB_TOKEN']
+            ).prefetch_related('organizations').get()
+        except Integration.DoesNotExist:
+            logger.info(
+                'gitlab.webhook.invalid-organization',
                 extra={
-                    'organization_id': organization_id,
+                    'external_id': request.META['HTTP_X_GITLAB_TOKEN'],
                 }
             )
             return HttpResponse(status=400)
 
-        body = six.binary_type(request.body)
-        if not body:
-            logger.error(
-                PROVIDER_NAME + '.webhook.missing-body', extra={
-                    'organization_id': organization.id,
+        if integration.organizations.count() != 1:
+            logger.info(
+                'gitlab.webhook.extra-organizations',
+                extra={
+                    'count': len(integration.organizations),
+                    'external_id': integration.external_id,
                 }
             )
             return HttpResponse(status=400)
 
         try:
-            handler = self.get_handler(request.META['HTTP_X_EVENT_KEY'])
-        except KeyError:
-            logger.error(
-                PROVIDER_NAME + '.webhook.missing-event', extra={
-                    'organization_id': organization.id,
+            event = json.loads(request.body.decode('utf-8'))
+        except JSONDecodeError:
+            logger.info(
+                'gitlab.webhook.invalid-json',
+                extra={
+                    'external_id': integration.external_id
                 }
             )
             return HttpResponse(status=400)
 
-        if not handler:
-            return HttpResponse(status=204)
-
         try:
-            event = json.loads(body.decode('utf-8'))
-        except JSONDecodeError:
-            logger.error(
-                PROVIDER_NAME + '.webhook.invalid-json',
-                extra={
-                    'organization_id': organization.id,
-                },
-                exc_info=True
-            )
+            handler = self._handlers[request.META['HTTP_X_GITLAB_EVENT']]
+        except KeyError:
+            logger.info('gitlab.webhook.missing-event', extra={
+                'event': request.META['HTTP_X_GITLAB_EVENT']
+            })
             return HttpResponse(status=400)
 
-        handler()(organization, event)
+        handler()(integration, integration.organizations.first(), event)
```

Comment:
Why only pick the first organization associated with the integration to create the `PullRequest` or `Commit` objects? That is going to run into a couple of issues:
- Because GitLab uses default identity (Oauth only), all the auth information is supplied by the user. If, for some reason, the default auth in the first organization doesn't work (the user deletes their account or something of that nature), this will end the attempt to update. But if you had tried the second organization's auth  it would have succeeded, you fail to update the information altogether.
- Second, if there are other organizations associated with the integration, you are not updating them, as both `PullRequest` and `Commit` model objects are scoped via an `organization_id`. You will need to create those objects for all orgs associated with an integration for the behavior to work as expected.

Reply:
> Second, if there are other organizations associated with the integration

Is this possible? I observed locally that each organization that connected the gitlab integration made a new `sentry_integration` record in the database. Each of these records only ever had 1 record in the join table. I'm assuming the ManyToMany association is either for other integrations or old behavior.

Reply:
Yes it is possible and expected. I would suggest not looking at Github as the example, in many ways it is a special case. In GitHub Apps there is no way to have a many to many relationship. This is a restraint on their end, but not on ours. 

The many to many relationship is intended. The idea is that when you install the same account to a different organization all that has to be added is another orangizationIntegration model. 
And the integration itself can be shared across organizations that way.

So for all other integrations but GitHub, the many to many relationship is expected behavior and must be handled here. The closest integration to GitLab is AzureDevOps due to the default auth. You will run into very similar issues I did when creating that integration. But they are issues that must be handled.

Reply:
> The idea is that when you install the same account to a different organization all that has to be added is another orangizationIntegration model. And the integration itself can be shared across organizations that way.

I installed the current GitLab integration into different sentry organizations as the same user and as different ones, and each install created new `sentry_integration` records. I wasn't able to reproduce an integration being shared by multiple organizations. The OrganizationIntegrationDetails deletion logic doesn't look like it would correctly handle integrations being shared across organizations either.

https://github.com/getsentry/sentry/blob/d125ced4f403086d03a6994a4b43d534827b9c4c/src/sentry/api/endpoints/organization_integration_details.py#L30-L56

A delete in one organization would delete the integration from all the other organizations as well.

Reply:
I was able to find multiple organizationintegrations for slack, jira and bitbucket. I don't understand how that state is created, but I'll update the gitlab implementation so that if does happen it behaves correctly.